earthfile2llb/converter.go

```diff
@@ -626,10 +648,15 @@ func (c *Converter) RunExpression(ctx context.Context, expressionName string, op
 		}
 	}
 
-	// Perform execution, but append the command with the right shell incantation that
-	// causes it to output to a file. This is done via the shellWrap.
-	opts.shellWrap = withShellAndEnvVarsOutput(outputFile)
-	opts.WithShell = true // force shell wrapping
+	// Redirect stdout to an output file *before* running any commands.
+	// This should be done rather than appending ">output" to the command, in order to maintain
+	// compatibility with /bin/sh's shelling-out, e.g. consider the command:
+	//   echo "hello $(echo -n one && echo -n two >/dev/null && echo -n three)."
+	// it will produce
+	//   "hello onethree."
+	// however, if we appended ">output", instead earthly would errornously output "hello three."
+	opts.Args = []string{fmt.Sprintf("exec 1<>'%s' && %s", escapeShellSingleQuotes(outputFile), opts.Args[0])}
```

Comment:
I initially tried moving the `echo` prepend to occur higher up, because in my mind once we hit a shell wrap, I expect the args to to be a valid command.

I could try adding a new bool such as `isExpression`, or `requiresEcho` to

```
func strWithEnvVarsAndDocker(args []string, envVars []string, withShell, withDebugger, forceDebugger, withDocker bool, exitCodeFile string, isExpression bool) string {
```

which would prepend the `echo`.

alternatively, we could introduce a `command string` arg, which would turn into:

```
func strWithEnvVarsAndDocker(command string, args []string, envVars []string, withShell, withDebugger, forceDebugger, withDocker bool, exitCodeFile string) string {
```

but that might make the cases where it is a valid command more verbose as we would then have to do `strWithEnvVarsAndDocker(args[0], args[1:])` along with `if len(args) == 0 return error` handling.

I'm leaning towards `isExpression`, does that sound reasonable?

Reply:
similarly when an output file is passed, i can prepend the `"exec 1<>'%s' && ` bit

Reply:
Yeah, `isExpression` sounds good. We could maybe also put the args to that into some kind of opt? Perhaps in a different PR.